sos/plugins/juju.py

```diff
@@ -9,113 +9,201 @@
     ]
 
-    def get_deployed_services(self):
-        cmd = "juju status --format json"
-        status_json = self.call_ext_prog(cmd)['output']
-        self.add_string_as_file(status_json, "juju_status_json")
-        # if status_json isn't in JSON format (i.e. 'juju' command not found),
-        # or if it does not contain 'services' key, return empty list
+    # Get the configuration of each application deployed in a model
+    def collect_app_config(self, username, modelname, statusfilespath):
+        statusinfo = {}
         try:
-            return json_loads(status_json)['services'].keys()
-        except ValueError:
-            return []
-
-    @ensure_service_is_running("juju-db")
-    def export_mongodb(self):
-        collections = (
-            "relations",
-            "environments",
-            "linkednetworks",
-            "system",
-            "settings",
+            fp = open(statusfilespath, "r")
+            statusinfo = json_loads(fp.read())
+            fp.close()
+        except Exception:
+            return
+        if "applications" in statusinfo:
+            for appname, appinfo in statusinfo["applications"].items():
+                # Get the application config for every application in every
+                # model
+                self.add_cmd_output(
+                    "sudo -i -u {} juju config --format yaml -m {} {}".format(
+                        username, modelname, appname
+                    ),
+                    suggest_filename="juju_config_-m_{}_{}.yaml".format(
+                        modelname, appname
+                    ),
+                )
+
+    # Get some information pertaining to each model
+    def collect_model_output(self, username, modelsfilepath):
+        modelsinfo = {}
+        try:
+            fp = open(modelsfilepath, "r")
+            modelsinfo = json_loads(fp.read())
+            fp.close()
+        except Exception:
+            return
+        if "models" in modelsinfo:
+            for model in modelsinfo["models"]:
+                if "short-name" in model:
+                    modelname = model["short-name"]
+                    # Run these commands for each model
+                    model_cmds = [
+                        "juju debug-log -n 500 --no-tail -m " + modelname,
+                        "juju model-config --format yaml -m " + modelname,
+                    ]
+                    for cmd in model_cmds:
+                        # The file names can be long and confusing from the
+                        # juju commands, clean them up and give them an
+                        # appropriate extension if possible
+                        filename = self._mangle_command(cmd)
+                        if re.match(r".*_--format_yaml.*", filename):
+                            filename = re.sub(r"_--format_yaml", "", filename)
+                            filename += ".yaml"
+                        self.add_cmd_output(
```

Comment:
What's the point of all this complexity? Why is it necessary or useful to strip out `_--format_yaml`? This seems like a very odd way to treat the command file names.

Reply:
(It is also adding, like, 12 lines to the plugin, which could otherwise be handled in a single `add_cmd_output([])` call).